### ISSUE TITLE
PuLID for Flux released, hope to see implementation on ComfyUI

### ISSUE THREAD

**Issue by Orenji-Tangerine** [cubiq/PuLID_ComfyUI#69]
![image](https://github.com/user-attachments/assets/a31c0fa3-7df1-4319-8cbf-21ce8ca2e94f)

https://github.com/ToTheBeginning/PuLID/blob/main/docs/pulid_for_flux.md

**Comment by defertoexpertise** [comment]
Was just coming to ask if the original code will work with the new model or if adaptors will be needed.

**Comment by cubiq** [comment]
no we need a new code. I'll check today if it's feasible

**Comment by cchance27** [comment]
I started trying to port it but not great with the code lol, the model layer names changed that’s for sure and there’s different embedders for flux

in the test code I’m trying I got it to run and diffuse… but the likeness doesn’t port over so I broke something LOL

does EVA2 always throw warnings about incompatible layers or is that something I did

**Comment by cubiq** [comment]
yeah I believe eva always returned those errors

**Comment by defertoexpertise** [comment]
OK so thats not what i broke :) I broke something... else...

**Comment by lschaupp** [comment]
> no we need a new code. I'll check today if it's feasible

any update?

**Comment by Rosicky7Arsenal** [comment]
PuLID for Flux^_^

**Comment by Spritan** [comment]
Hoping for good news

**Comment by cubiq** [comment]
I'm looking into it

**Comment by AoeSyL** [comment]
i need puLID for flux...

**Comment by defertoexpertise** [comment]
any chance you need help, didnt see any branches for flux, i had it working but sorta trashed my work lol, cause the patching wasn't working correctly, i had the idformer working but the second block wasn't working the part that used to be XY from SD verison.

**Comment by cubiq** [comment]
it's a bit more complicated than I'd hoped, so it will take some more time as it's not my top priority now, sorry guys. the model is still not perfect anyway so you are not missing too much :stuck_out_tongue:

**Comment by yincangshiwei** [comment]
You can choose ComfyUI_StoryDiffusion, which already supports PuLID.

**Comment by cubiq** [comment]
ComfyUI_StoryDiffusion is a diffusers sandbox

**Comment by sh-alim** [comment]
I still couldn't get ComfyUI_StoryDiffusion to work with Flox. RTX4090/14900K/32GB are not enough for it

**Comment by cchance27** [comment]
I'm busy working on getting flux working with pulid... as it's my first time working on patching the model, and i'm definitly doing it wrong, and my tensors are all over the place, and many issues to fix (probably will take longer than cubiq getting his verison done)... i can say it "works"

<img width="1131" alt="image" src="https://github.com/user-attachments/assets/dcbef6ec-1f1d-4422-954e-bcbc08aed4d9">

Need to get it cleaned up and figure out some major bugs (the patching is happening on the original model and not the cloned model in my code i wrote lol, so adding a different ksampler pulling from the model node and not the pulid node breaks everything lol, and i've got most of my devices and tensors hardcoded for mac but... i mean... it works LOL And this is from GGUF/Q8... havent tested it with standard checkpoint yet.

**Comment by cubiq** [comment]
@cchance27 please let's not do the work twice, show your code and talk about it, or contact me on discord.

at the moment the only way I've found to do it without a dedicated ksampler is to monkey patch the flux pipeline, even though they are releasing new hooks for comfyui that will make things easier.

**Comment by cczw2010** [comment]
> ComfyUI_StoryDiffusion

ComfyUI_StoryDiffusion is supported now：

> if using pulid-flux,No need to enter 'cpu' in easyfunction. Now choose CPU offloading based on your VRAM, with the dividing points being VR>30G, 18G<VR<30G, or VR<18G;

**Comment by chimera-creator** [comment]
> I'm busy working on getting flux working with pulid... as it's my first time working on patching the model, and i'm definitly doing it wrong, and my tensors are all over the place, and many issues to fix (probably will take longer than cubiq getting his version done)... i can say it at a minimum it is possible XD and "works"
> 
> <img alt="image" width="1131" src="https://private-user-images.githubusercontent.com/13766359/371780694-dcbef6ec-1f1d-4422-954e-bcbc08aed4d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc2MjM3MzUsIm5iZiI6MTcyNzYyMzQzNSwicGF0aCI6Ii8xMzc2NjM1OS8zNzE3ODA2OTQtZGNiZWY2ZWMtMWYxZC00NDIyLTk1NGUtYmNiYzA4YWVkNGQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI5VDE1MjM1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjNDBhMTgyNjk2MmRhODk2MmI3NzFiYTZiZjE2YThhMThkY2QxMTAyN2JlNGYyZjdjNGNkNGU0YTc1MDM2MWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.89thgFNqlWF6RJfvl3d2jJYNcewyEpzlGLcuQUPQ3DE">
> Need to get it cleaned up and figure out some major bugs (the patching is happening on the original model and not the cloned model in my code i wrote lol, so adding a different ksampler pulling from the model node and not the pulid node breaks everything lol, and i've got most of my devices and tensors hardcoded for mac but... i mean... it works LOL And this is from GGUF/Q8... havent tested it with standard checkpoint yet.
> 
> I haven't played with pulid enough myself, so not sure if the over-fitting to the image is a weight issue or a pulid issue or a bug i made in my code, will find out later got work now.

I'd love to play around with the code. :)

**Comment by adamreading** [comment]
> cpu

Spent about 8 hours trying to get it running - but it’s not quite there yet. Hopefully though once the bugs are fixed…

**Comment by chimera-creator** [comment]
> > cpu
> 
> Spent about 8 hours trying to get ComfyUI_StoryDiffusion running - but it’s not quite there yet. Hopefully though once the bugs are fixed…

Same. When you go to the ComfyUI_StoryDiffusion github lots of people were having issues with the implementation the day this post was made. It also doesn't support GGUF which is what a lot of folks are using only full and NF4.

**Comment by adamreading** [comment]
At the moment - once it doesn’t crash. It only works with 2 specific diffuser versions - but I get errors before it gets to output - and the creator is on holidays for a few days

**Comment by chimera-creator** [comment]
> At the moment - once it doesn’t crash. It only works with 2 specific diffuser versions - but I get errors before it gets to output - and the creator is on holidays for a few days

Got it working. It's EXTREMELY slow (rtx4090) but works. My error was regarding antelope for face analysis. The auto download doesn't download to the model directory. It tries to save to a root directory instead. Even worse than that when it unzips the model it its a folder in a folder. You have to flatten the directory.

**Comment by cchance27** [comment]
https://github.com/cchance27/ComfyUI-PuLID4Flux here's my horrible implementation XD it works on my macbook m3 but like i said many times in the readme, its really not done the right way but don't know the comfy backend well enough yet and like cubiq said, comfy itself def seems to be lacking the hooks we'd need to modify the block forwards

**Comment by cubiq** [comment]
yeah that's monkey patching, which is what we should not do :smile: also you use deepcopy, probably it works on a mac because of the shared memory but I get OOM on my machine with it... so cloning, patching, unpatching is a bit more complicated.

**Comment by cchance27** [comment]
Ya i figured as much, but i figured worth sharing as proof it does work lol, hopefully we'll get some form of way to add a step in the model/block forwards.. as that would make things a LOT easier.

As for the deepcopy, i used that code basically straight from the original modelpatcher code :S

**Comment by cubiq** [comment]
totally worth sharing, thanks! That is where I am at the moment too, like I said without monkey patching I can't figure out how to do it. I asked the comfyorg and they confirmed they are working on the right hooks to use for Flux, so I don't know if it's worth releasing the version that I have now or wait for the updated comfyui.

> As for the deepcopy, i used that code basically straight from the original modelpatcher code :S

yeah it's fine but that way you are basically duplicating everything twice. I haven't checked, do you clone the original model? Flux is extremely heavy and duplicating everything gives me OOM real soon. Without it the problem is that you are patching the original model and you basically need an "unpatch" node.

**Comment by balazik** [comment]
Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: [https://github.com/balazik/ComfyUI-PuLID-Flux](https://github.com/balazik/ComfyUI-PuLID-Flux)
Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).

**Comment by cchance27** [comment]
Really cool, love that its documented! And you definitely got further than i did :)

also gotta say i feel this comment in my soul... lol
```
# Patch the Flux model (original diffusion_model)
# Nah, I don't care for the official ModelPatcher because it's undocumented!
# I want the end result now, and I don’t mind if I break other custom nodes in the process. 😄
```

I'm actually going to refactor out my version from the modelpatcher as it isn't actually giving me what i wanted it to as its still breaking the original model the way i did it so i might as well avoid the copy and just directly modify the original model object.

**Comment by balazik** [comment]
Oh Yes! It wasn't just a whim decision but lack of knowledge of how else to patch the model without breaking something else. I also tried to run attention masking but I didn't manage to do that, plus I still have to understand how exactly Flux generates the "img" with padding because then I just need to multiply the img with mask directly in the forward_orig function (or at least I hope so).

**Comment by cchance27** [comment]
Whats funny is looking at the code from the original forward, if they had the proper hooks, even controlnet could move out of that function as it technically shouldn't really be in that forward, it should be just another hook if a controlnet is needed... but because they don't have a hook they have to add the hardcoded check to look for controlnet.

Sadly now when they add the hook they have to leave controlnet check in their because removing it will break controlnet for people with the old nodes.

**Comment by cubiq** [comment]
just FYI this seems to be the branch with the hooks https://github.com/comfyanonymous/ComfyUI/tree/patch_hooks

**Comment by enzoliao** [comment]
> Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).

Just tried and it works for me...

**Comment by adamreading** [comment]
> Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).

This is so cool, I have it running in the cloud at about 12-15 seconds a generation (2nd onwards). Thanks for all your help and jumping onto the cloud instance to see what was going on!!

**Comment by Orenji-Tangerine** [comment]
> Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).

![ComfyUI_temp_qmfkq_00002_](https://github.com/user-attachments/assets/bf07807b-93d6-400f-a49e-af6a8e82eafc)

![image](https://github.com/user-attachments/assets/9bb5cfbe-aa2c-4b72-a213-40492807a186)
Thanks! Got it working, first run 71 sec, second run 24 sec on a 4060Ti 16GB on the GGUF model (Pytorch 2.1.2 + CU 11.8)

![image](https://github.com/user-attachments/assets/ea7540f6-fa77-4b7f-aa01-aed64ba0241e)
FP8 one is faster on my system, first run 48 sec, second run 21 sec

**Comment by adamreading** [comment]
> > Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).
> 
> This is so cool, I have it running in the cloud at about 12-15 seconds a generation (2nd onwards). Thanks for all your help and jumping onto the cloud instance to see what was going on!!

I now it have it running Flux unchained (fp8) and my own character Lora - it takes 10 minutes to load (i think that’s a Ram issue) but only 17 seconds a generation once loaded (24Vram 16 Ram) - I’ll test it on the 32Mb Ram later and see if it speeds up - and that also includes a 2x upscale with a model.

**Comment by sipie800** [comment]
> Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).

Great work. What do you mean by not supporting attention mask, is official pulid with that feature, what's it for?

**Comment by Orenji-Tangerine** [comment]
> > Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).
> 
> Great work. What do you mean by not supporting attention mask, is official pulid with that feature? What's it for?

![ComfyUI_temp_dzilf_00001_](https://github.com/user-attachments/assets/923f0ff4-b6c5-46fe-955c-6ed36523a774)
![ComfyUI_temp_hfkmk_00001_](https://github.com/user-attachments/assets/151c4d92-daf0-4e5b-a08b-25c0e7da6626)

It should be this, so that each masked area can be applied with PuLID with respective photo input

**Comment by Orenji-Tangerine** [comment]
> > Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).
> 
> Great work. What do you mean by not supporting attention mask, is official pulid with that feature? What's it for?

![image](https://github.com/user-attachments/assets/01998636-0361-4c24-98b1-272d1d6401d0)

It should be this, so that each masked area can be applied with PuLID with respective photo input

**Comment by fideuacaprichos** [comment]
Are you guys going to keep talking about ComfyUI-PuLID-Flux here forever? Isn't it tedious to check both repositories?

**Comment by chimera-creator** [comment]
> Hi all. I also just finished a PuLID node for ComfyUI Flux last night. Here: https://github.com/balazik/ComfyUI-PuLID-Flux Also a proof of concept (alpha version). I completely skipped the ModelPatcher because I wanted a result. It doesn't support attention masks yet, but it can run as a chain (so you can combine face embeddings). It should run with all Flux.1-dev models (16bit, GGUF, FP8 e5m2 etc).

Awesome work, dude! I can generate like 200 image sin the time it took me to do one before. Results look great to me with an upscale.
![TUS16-50-32_00001_](https://github.com/user-attachments/assets/1464ea58-b06e-4dd2-bf30-fbf230b177bf)

**Comment by lschaupp** [comment]
any update on this

**Comment by xps2020** [comment]
Any update about mask function？
